File: backends/java/zappy-domain/src/main/java/com/zappymart/domain/promotions/PromotionRule.java
package com.zappymart.domain.promotions;

import com.zappymart.domain.shared.Money;

public sealed interface PromotionRule {

    PromotionKind kind();

    Money discountFor(Money subtotal);

    boolean carriesShipping();

    record PercentageOff(int percentage) implements PromotionRule {
        public PercentageOff {
            if (percentage < 1 || percentage > 100) {
                throw new IllegalArgumentException("A percentage code takes between 1 and 100 percent: " + percentage);
            }
        }

        @Override
        public PromotionKind kind() {
            return PromotionKind.PERCENTAGE;
        }

        @Override
        public Money discountFor(Money subtotal) {
            return subtotal.percentageRoundedHalfUp(percentage);
        }

        @Override
        public boolean carriesShipping() {
            return false;
        }
    }

    record FixedAmountOff(Money amount) implements PromotionRule {
        public FixedAmountOff {
            if (amount.isZero()) {
                throw new IllegalArgumentException("A fixed amount code takes more than nothing off");
            }
        }

        @Override
        public PromotionKind kind() {
            return PromotionKind.FIXED_AMOUNT;
        }

        @Override
        public Money discountFor(Money subtotal) {
            return amount.cappedAt(subtotal);
        }

        @Override
        public boolean carriesShipping() {
            return false;
        }
    }

    record FreeShipping() implements PromotionRule {
        @Override
        public PromotionKind kind() {
            return PromotionKind.FREE_SHIPPING;
        }

        @Override
        public Money discountFor(Money subtotal) {
            return Money.zero();
        }

        @Override
        public boolean carriesShipping() {
            return true;
        }
    }
}

File: backends/kotlin/zappy-domain/src/main/kotlin/nl/zappymart/domain/promotions/PromotionRule.kt
package nl.zappymart.domain.promotions

import nl.zappymart.domain.shared.Money

sealed interface PromotionRule {

    val kind: PromotionKind

    fun discountFor(subtotal: Money): Money

    data class Percentage(val percentage: Int) : PromotionRule {

        init {
            require(percentage in 1..100) { "A percentage code takes between 1 and 100 percent, not $percentage" }
        }

        override val kind = PromotionKind.PERCENTAGE

        override fun discountFor(subtotal: Money) = subtotal.percentageRoundedHalfUp(percentage)
    }

    data class FixedAmount(val amount: Money) : PromotionRule {

        override val kind = PromotionKind.FIXED_AMOUNT

        override fun discountFor(subtotal: Money) = amount.cappedAt(subtotal)
    }

    data object FreeShipping : PromotionRule {

        override val kind = PromotionKind.FREE_SHIPPING

        override fun discountFor(subtotal: Money) = Money(0, subtotal.currency)
    }
}
